Source/Orts.Simulation/Simulation/Physics/Train.cs

```diff
@@ -9257,12 +9318,56 @@ public void UpdateExplorerMode(int signalObjectIndex)
 
                 if (endAuthority == END_AUTHORITY.SIGNAL)
                 {
-                    TrackCircuitSection lastSection = signalRef.TrackCircuitList[newRoute[newRoute.Count - 1].TCSectionIndex];
-                    int lastDirection = newRoute[newRoute.Count - 1].Direction;
-                    if (lastSection.EndSignals[lastDirection] != null && lastSection.EndSignals[lastDirection].thisRef == nextUnclearSignalIndex)
+                    // The logic here is to keep the value of SNCA of first signal found in path, and reduce this value as cleared signals are passed.
+                    // When an uncleared signal is found, it is requested to clear if SNCA of first signal is not satisfied.
+                    // If we are far away from the first signal, there is no point in clearing it until we get closer.
+                    if (unclearedSignal && signalIndex < newRoute.Count)
                     {
                         SignalObject reqSignal = signalRef.SignalObjects[nextUnclearSignalIndex];
-                        newRoute = reqSignal.requestClearSignalExplorer(newRoute, forward ? routedForward : routedBackward, false, 0);
+                        bool firstSignalPassed = false;
+                        int numCleared = 0;
+                        totalLengthM = 0;
+                        offsetM = direction == 0 ? requiredPosition.TCOffset : thisSection.Length - requiredPosition.TCOffset;
+                        for (int iindex = 0; iindex < newRoute.Count && (firstSignalPassed || totalLengthM < minCheckDistanceExplorerM) && (!firstSignalPassed || numCleared != 0); iindex++)
+                        {
```

Comment:
```suggestion
                        for (int iindex = 0; iindex < newRoute.Count; iindex++)
                        {
                            var continue1 = firstSignalPassed || totalLengthM < minCheckDistanceExplorerM;
                            var continue2 = !firstSignalPassed || numCleared != 0;
                            if (!continue1 || !continue2)
                                break;
```
That's a lot of booleans to keep track of, there; please consider this style instead, with more descriptive names for "continue1" and "continue2".